x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -354,6 +369,48 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
         }
     }
 
+    private final Map<IndicesAccessControlCacheKey, ListenableFuture<IndicesAccessControl>> indicesAccessControlLookup =
+        new ConcurrentHashMap<>();
+
+    private void buildIndicesAccessControlAsync(
+        String action,
+        AuthorizationInfo authorizationInfo,
+        ResolvedIndices resolvedIndices,
+        Metadata metadata,
+        ActionListener<IndexAuthorizationResult> listener) {
+
+        final ActionListener<IndicesAccessControl> wrappingListener =
+            listener.map((IndicesAccessControl indicesAccessControl) -> new IndexAuthorizationResult(true, indicesAccessControl));
+
+        if (authorizationInfo instanceof RBACAuthorizationInfo) {
+            final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
+            final Tuple<IndicesPermission, IndicesPermission> indicesPermissionTuple = getRoleIndicesPermissions(role);
+            final Set<String> requestedIndices = Set.copyOf(resolvedIndices.getLocal());
+            final IndicesAccessControlCacheKey cacheKey = new IndicesAccessControlCacheKey(metadata.version(),
+                action, requestedIndices,
+                indicesPermissionTuple.v1(),
+                indicesPermissionTuple.v2());
```

Comment:
It looks to me like a lot of code to compute a cache key; it's hard to follow and counter-productive to the performance goal of this all.

I think we should work with the fact that the `Role`, `resolvedIndices.getLocal`, `action` and `metadata.getIndicesLookup()` are immutable (or effectively immutable, maybe assert or test it).
In this case, the cache key class's `equals` method would use the `==` test for the corresponding members; analogous with `System#identityHashCode`.

Reply:
I don't quite follow. Maybe I am missing something, but how is it possible to use `==` to compare `action` or `resolvedIndices.getLocal`? For example, even if two `action` strings have the same value, they are different objects belong to two different HTTP worker threads? 

Comparing identity of `metadata.getIndicesLookup()` may be viable, but it does not seem to be more convenient or more reliable than comparing a simple version number?

As for `Role`, the issue is that the roles for API keys are [constructed on the fly](https://github.com/elastic/elasticsearch/blob/eeb09f0248765c7b24a4ba8129e29141e1e7c956/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L302) for each request. So even the same API key will get two equavilent `Role` objects but with different identities. To get around it, the cacheKey here uses only the `IndicesPermission` component of a `Role` and the comparison for `IndicesPermission` is an object identity.